feature/text_based.py

```diff
@@ -0,0 +1,703 @@
+        column corresponds to a feature.
+        The version of code for generating labels from categories has been moved to utils.py.
+
+        Parameters
+        ----------
+        input_df: pd.DataFrame
+            Input data frame with categories and features of content to select from.
+
+        categories: pd.DataFrame
+            Data frame that contains categories/labels to be covered.
+
+        selection_size: int
+            Number of feature to select.
+
+        optimization_method: str
+            Code prevents performing featurization for the case mentioned in 'option_map' dictionary with
+            random, greedy and exact optimization.
+
+        cost_metric: str
+            Code prevents performing featurization for the case mentioned in 'option_map' dictionary with
+            unicost or diverse cost metric.
+
+        Returns
+        -------
+        No return
+       --------
+        """
+        self.matrix = categories.to_numpy()
+        self._num_rows, self._num_cols = self.matrix.shape
+
+        # option_map contains the tuple of optimization_method, selection_size, and cost_metric
+        # that do not need featurization method
+        option_map = {
+            ("random", True, "diverse"): lambda: None,
+            ("random", False, "unicost"): lambda: None,
+            ("random", True, "unicost"): lambda: None,
+            ("greedy", False, "unicost"): lambda: None,
+            ("greedy", True, "unicost"): lambda: None,
+            ("exact", True, "unicost"): lambda: None,
+            ("exact", False, "unicost"): lambda: None
+        }
+
+        # If the entry of in the dictionary matches below argument set, then corresponding lambda function is retrieved
+        options = option_map.get((optimization_method, selection_size is not None, cost_metric))
+        if options:
+            options()
+        else:
+            feature_column = self.featurization_method.fit_transform(input_df)
+            self.features = np.array([eval(txt_feature) if isinstance(txt_feature, str)
+                                      else txt_feature for txt_feature in feature_column.tolist()])
+            if len(self.features) != self._num_cols:
+                raise ValueError(f"Process Data Error: features size ({len(self.features)}) should match the number"
+                                 f"of columns ({self._num_cols})")
+        if self.matrix.ndim != 2:
+            raise ValueError("Process Data Error: matrix should 2D")
+        if self.selection_size is not None and self.selection_size > self._num_cols:
+            raise ValueError("Process Data Error: selection_size cannot exceed num columns")
+
+    def _select_multi_level_optimization(self) -> List:
+
+        unicost = np.ones(self._num_cols)
+        data = Data(cost=unicost, matrix=self.matrix)
+
+        if self.selection_size is None:
+            if self.cost_metric == "unicost":
+                selected = self._solve_set_cover(data)
```

Comment:
don't you need to set k here ?

Reply:
No, we do not need it. 'k' is only required when num_features is not None. For these two conditions

Punicost=unicost().​
If t <unicost_selection,
max_cover(Punicost, t)

AND

k=unicost ().​
diverse_costs=kmeans(k). ​
P_diverse=diverse().  ​
If t <diverse_selection, 
max_cover(P_diverse, t).